DTensor: add more foreach ops to supported sharding prop list

bdhirsh opened the pull request:
fixes https://github.com/pytorch/pytorch/issues/132016. 

Right now if you run an op that DTensor has no sharding prop rule, **and** that op accepts non-trivial pytrees of inputs tensors as arguments, DTensor can end up infinite looping before it has the chance to error due to not having a sharding prop rule.

This PR doesn't fix the problem, but adds rules for the culprit ops (missing foreach ops)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131481
* __->__ #132066



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

bdhirsh committed:
DTensor: use pytrees to convert DTensors into specs

[ghstack-poisoned]

bdhirsh commented on torch/distributed/_tensor/_dispatch.py:
```diff
@@ -374,7 +384,6 @@ def try_get_replicate_spec(
             else:
                 kwargs_schema[k] = v
                 local_kwargs[k] = v
-
```
before landing this, I probably need to:

(1) try adding a sharding prop rule for `aten._foreach_mul`
(2) see if that fixes the repro and add tests for it

bdhirsh committed:
Update on "DTensor: use pytrees to convert DTensors into specs"

I think this partially fixes https://github.com/pytorch/pytorch/issues/132016. It doesn't fully fix it though because:

(1) we don't have a sharding prop rule for `aten. _foreach_mul_`
(2) when we don't have a sharding prop rule, we assume that the schema of our op just has a flat list of inputs that does not require any nested unflattening




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wz337 commented on torch/distributed/_tensor/ops/_pointwise_ops.py:
```diff
@@ -546,9 +546,15 @@ def linear_pointwise_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> Strategy
     aten._foreach_clamp_min_.Scalar,
     aten._foreach_div_.List,
```
Should we also add aten. _foreach_div_.Scalar and aten. _foreach_div.Scalar in this PR as well?

bdhirsh replied:
yep good call

wanchaol commented on torch/distributed/_tensor/_dispatch.py:
```diff
@@ -356,6 +364,8 @@ def try_get_replicate_spec(
                 args_schema.append(arg)
                 local_args.append(arg)
 
+        tree_map(arg_to_spec, args)
```
hmm trying to understand more on this, we do have pytree to flatten the args input for certain ops, i.e. the foreach op list have pytree flatten enabled https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/ops/_pointwise_ops.py#L641

bdhirsh replied:
oh right, is the idea that DTensor wants to manually specify which ops actually need the pytree machinery, so their usage is more limited (to avoid the perf impact?)

In that case, we can probably tweak this code to only do the tree_map here for ops that have opted into "requiring pytrees" (probably by checking the flag that you just linked)

wanchaol replied:
> oh right, is the idea that DTensor wants to manually specify which ops actually need the pytree machinery, so their usage is more limited (to avoid the perf impact?)

Yep!

> In that case, we can probably tweak this code to only do the tree_map here for ops that have opted into "requiring pytrees" (probably by checking the flag that you just linked)

We already sorta doing that in a few lines above https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291-L294

So I am wondering if there're some additional bugs that triggered that issue

bdhirsh replied:
oh you are totally right... let me take another look

bdhirsh replied:
ah yep you're right. So:

(1) this logic does the right thing of operating on the flattened args, as long as we detect that the op's sharding rule has opted into using pytrees

(2) the only bug is that DTensor was missing sharding rules for a few foreach ops, causing DTensor to use the "default" path of not using pytrees, in a "bad" way that causes DTensor to infinite loop horrible (I think, before getting a chance to error about not having a proper sharding rule).

hmm @wanchaol - is there any easy way to know earlier in advance if an op has no sharding prop rule and thus will fail, so we can check it here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291

One option is maybe to detect when there's no sharding prop rule and error earlier, before we try to generate FakeTensor arguments.

bdhirsh replied:
(for now, I just "fixed" the issue by updated the foreach_ops list that DTensor relies on)

wz337 replied:
> ah yep you're right. So:
> 
> (1) this logic does the right thing of operating on the flattened args, as long as we detect that the op's sharding rule has opted into using pytrees
> 
> (2) the only bug is that DTensor was missing sharding rules for a few foreach ops, causing DTensor to use the "default" path of not using pytrees, in a "bad" way that causes DTensor to infinite loop horrible (I think, before getting a chance to error about not having a proper sharding rule).
> 
> hmm @wanchaol - is there any easy way to know earlier in advance if an op has no sharding prop rule and thus will fail, so we can check it here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291
> 
> One option is maybe to detect when there's no sharding prop rule and error earlier, before we try to generate FakeTensor arguments.

For (2), I think the issue is that `_propagate_tensor_meta(op_schema)` get called first and errors out so we could see all different errors (here are some more examples where we don't see the infinite loop but some other error: https://github.com/pytorch/pytorch/issues/124990) even though the true reason is that no shardping prop has been registered for an op.   
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_sharding_prop.py#L198

bdhirsh committed:
Update on "DTensor: use pytrees to convert DTensors into specs"

I think this partially fixes https://github.com/pytorch/pytorch/issues/132016. It doesn't fully fix it though because:

(1) we don't have a sharding prop rule for `aten. _foreach_mul_`
(2) when we don't have a sharding prop rule, we assume that the schema of our op just has a flat list of inputs that does not require any nested unflattening

UPDATE: the linked repro passes now that I added the missing foreach overloads, so DTensor registers proper sharding rules for them.




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: use pytrees to convert DTensors into specs"

I think this partially fixes https://github.com/pytorch/pytorch/issues/132016. It doesn't fully fix it though because:

(1) we don't have a sharding prop rule for `aten. _foreach_mul_`
(2) when we don't have a sharding prop rule, we assume that the schema of our op just has a flat list of inputs that does not require any nested unflattening

UPDATE: the linked repro passes now that I added the missing foreach overloads, so DTensor registers proper sharding rules for them.




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm thanks!

wanchaol commented:
Please change the title and summary to reflect newest changes :)

bdhirsh committed:
Update on "DTensor: use pytrees to convert DTensors into specs"

I think this partially fixes https://github.com/pytorch/pytorch/issues/132016. It doesn't fully fix it though because:

(1) we don't have a sharding prop rule for `aten. _foreach_mul_`
(2) when we don't have a sharding prop rule, we assume that the schema of our op just has a flat list of inputs that does not require any nested unflattening

UPDATE: the linked repro passes now that I added the missing foreach overloads, so DTensor registers proper sharding rules for them.




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

awgu commented:
@pytorchbot merge
